File: PP/lista4/lista4.ml
let rec print_list = function
  | [] -> ()
  | h::t -> print_int h; print_string " "; print_list t;
;;

type 'a bt =
  | Empty
  | Node of 'a * 'a bt * 'a bt
;;

type 'a graph =
  | Graph of ('a -> 'a list)
;;

let rec f a = f a;;

let breadthBT tt =
  let rec inside tl =
    match tl with
      | [] -> []
      | [Empty] -> []
      | Empty::t -> inside t
      | Node(v, left, right)::t -> v::(inside (t @ (left::right::[])))
  in
    inside [tt]
;;


let lenIn tt =
  let rec inside tt a =
    match tt with
      | Empty -> 0
      | Node(v, left, right) -> a + (inside left (a + 1)) + (inside right (a + 1))
  in
    inside tt 0


let lenEx tt =
  let rec inside tt a =
    match tt with
      | Empty -> a
      | Node(v, left, right) -> (inside left (a + 1)) + (inside right (a + 1))
  in
    inside tt 0

let depthSearch (Graph fn) st =
  let rec inside vis que =
    match que with
      | [] -> []
      | h::t -> if List.mem h vis then inside vis t else h::(inside(h::vis) (fn h @ t))
  in
    inside [] [st]
;;

let tt = Node(1,
 Node(2,
 Node(4,
 Empty,
Empty
 ),
 Empty
 ),
 Node(3,
 Node(5,
 Empty,
Node(6,
 Empty,
Empty
 )
 ),
 Empty
 )
 );;

print_list (breadthBT tt);;
print_string "\n";;
print_int (lenEx tt);;
print_string "\n";;
print_int (lenIn tt);;

File: PP/lista4/lista4.scala
sealed trait BT[+A]
  case object Empty extends BT[Nothing]
  case class Node[+A](elem:A, left:BT[A], right:BT[A]) extends BT[A]

sealed trait Graphs[A]
  case class Graph[A](succ: A=>List[A]) extends Graphs[A]


object lista4 {
  def printList(args: List[_]): Unit = {
    args.foreach(println)
  }

  def breadthBT[A](tt: BT[A]) = {
    def inside(tl: List[BT[A]]): List[A] = tl match {
        case Nil => Nil
        case Empty::t => inside(t)
        case Node(v, left, right)::t => v::(inside(t ++ List(left, right)))
    }
    inside(List(tt))
  }

  def lenIn[A](tt: BT[A]) = {
    def inside (tt:BT[A], a: Int): Int = tt match {
        case Empty => 0
        case Node(v, left, right) => a + (inside (left, (a + 1))) + (inside (right, (a + 1)))
    }
    inside(tt, 0)
  }

  def lenEx[A](tt: BT[A]) = {
    def inside (tt:BT[A], a: Int): Int = tt match {
        case Empty => a
        case Node(v, left, right) => (inside (left, (a + 1))) + (inside (right, (a + 1)))
    }
    inside(tt, 0)
  }


  def depthSearch [A](g: Graph[A])(st: A): List[A] = {
    def inside (vis: List[A])(que: List[A]): List[A] = que match {
        case Nil => Nil
        case h::t => if (vis contains h) inside(vis)(t) else h::inside(h::vis)((g succ h) ++ t)
    }
    inside (Nil)(List(st))
  }


  def main(args: Array[String]): Unit = {
    val tr =  Node(1,
 Node(2,
 Node(4,
 Empty,
Empty
 ),
 Empty
 ),
 Node(3,
 Node(5,
 Empty,
Node(6,
 Empty,
Empty
 )
 ),
 Empty
 )
 )
    printList(breadthBT(tr))
    println(lenIn(tr))
    println(lenEx(tr))
  }
}
